tools/compare.go

```diff
@@ -15,125 +15,291 @@
 
-	storeA, err := dstore.NewDBinStore(storeADef)
+	// Create a fileReader
+	fileReader, err := store.OpenObject(ctx, filename)
 	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeADef, err)
+		return nil, nil, fmt.Errorf("creating reader: %w", err)
 	}
 
-	storeB, err := dstore.NewDBinStore(storeBDef)
+	// Create a blockReader
+	blockReader, err := bstream.GetBlockReaderFactory.New(fileReader)
 	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeBDef, err)
+		return nil, nil, fmt.Errorf("creating block reader: %w", err)
 	}
 
-	blocksA := make(map[string]*pbeth.Block)
-	blocksB := make(map[string]*pbeth.Block)
-
-	filePrefix := fmt.Sprintf("%010d", blockNum)
-	collectBlocks := func(store dstore.Store, blockMap map[string]*pbeth.Block) error {
-		var files []string
-		err = storeA.Walk(ctx, filePrefix, func(filename string) (err error) {
-			files = append(files, filename)
-			return nil
-		})
-
-		var toClose []io.ReadCloser
-		defer func() {
-			for i := range toClose {
-				toClose[i].Close()
-			}
-		}()
+	var blockHashes []string
+	blocksMap := make(map[string]*pbeth.Block)
+	for {
+		curBlock, err := blockReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("reading blocks: %w", err)
+		}
 
-		for _, filepath := range files {
-			reader, err := store.OpenObject(ctx, filepath)
-			if err != nil {
-				return err
-			}
-			toClose = append(toClose, reader)
+		curBlockPB := curBlock.ToProtocol().(*pbeth.Block)
+		blockHashes = append(blockHashes, string(curBlockPB.Hash))
+		blocksMap[string(curBlockPB.Hash)] = curBlockPB
+	}
 
-			blockReader, err := bstream.GetBlockReaderFactory.New(reader)
-			if err != nil {
-				return err
-			}
+	return blockHashes, blocksMap, nil
+}
+
+func compareBlocksE(cmd *cobra.Command, args []string) error {
+	defer func() {
+		if err := recover(); err != nil {
+			fmt.Println(err)
+		}
+	}()
+	displayDiff := mustGetBool(cmd, "diff")
+
+	ctx := cmd.Context()
+	blockRange, err := bstream.ParseRange(args[2])
+	if err != nil {
+		return fmt.Errorf("parsing range: %w", err)
+	}
+	blockRangeSize, err := blockRange.Size()
+	if err != nil {
+		return fmt.Errorf("checking for valid range: %w", err)
+	}
+	if blockRangeSize == 0 {
+		return fmt.Errorf("invalid block range")
+	}
+
+	// Create stores
+	storeExpected, err := dstore.NewDBinStore(args[0])
+	if err != nil {
+		return fmt.Errorf("unable to create store at path %q: %w", args[0], err)
+	}
+	storeReceived, err := dstore.NewDBinStore(args[1])
+	if err != nil {
+		return fmt.Errorf("unable to create store at path %q: %w", args[1], err)
+	}
+
+	// Walk expected files
+	differentBlocks := make(map[string]block.Range)
+	blocksCountedInChunk := -1
+	chunkIsGood := true
+	rangeNum := 0
+	err = storeExpected.Walk(ctx, "00", func(filename string) (err error) {
```

Comment:
"00" is not always correct, is there is a block 160M so that the filename would be `0160000000`, you would match nothing.

The prefix should actually be dynamic, based on the start of the range, `sftools` has the logic to pick the right prefix.